x/metadata/keeper/query_server.go

```diff
@@ -73,10 +73,75 @@ func (k Keeper) GroupContext(c context.Context, req *types.GroupContextRequest)
 	return &types.GroupContextResponse{}, nil
 }
 
-// Record returns a collection of the records in a scope or a specific one by name
-func (k Keeper) Record(c context.Context, req *types.RecordRequest) (*types.RecordResponse, error) {
-	// TODO
-	return &types.RecordResponse{}, nil
+// RecordByScopeUUID returns a collection of the records in a scope or a specific one by name
+func (k Keeper) RecordByScopeUUID(c context.Context, req *types.RecordByScopeUUIDRequest) (*types.RecordByScopeUUIDResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+
+	if req.GetScopeUuid() == "" {
+		return nil, status.Error(codes.InvalidArgument, "scope uuid cannot be empty")
+	}
+
+	uuid, err := uuid.Parse(req.GetScopeUuid())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid scope uuid: %s", err.Error())
+	}
+
+	scopeAddr := types.ScopeMetadataAddress(uuid)
+	records, err := k.record(c, &scopeAddr, req.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	return &types.RecordByScopeUUIDResponse{ScopeUuid: req.GetScopeUuid(), ScopeId: scopeAddr.String(), Records: records}, nil
+}
+
+// RecordByScopeID returns a collection of the records in a scope or a specific one by name
+func (k Keeper) RecordByScopeID(c context.Context, req *types.RecordByScopeIDRequest) (*types.RecordByScopeIDResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+
+	if req.GetScopeId() == "" {
+		return nil, status.Error(codes.InvalidArgument, "scope id cannot be empty")
+	}
+
+	scopeAddr, err := types.MetadataAddressFromBech32(req.GetScopeId())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid scope id %s : %s", req.GetScopeId(), err)
+	}
+
+	scopeUUID, err := scopeAddr.ScopeUUID()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "unable to extract uuid from scope metaaddress %s", err)
+	}
+
+	records, err := k.record(c, &scopeAddr, req.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	return &types.RecordByScopeIDResponse{ScopeUuid: scopeUUID.String(), ScopeId: req.GetScopeId(), Records: records}, nil
+}
+
+// record returns a collection of the records in a scope or a specific one by name
+func (k Keeper) record(c context.Context, scopeAddress *types.MetadataAddress, name string) ([]*types.Record, error) {
```

Comment:
Should this func be capitalize: `Record` or is it private on purpose?

Reply:
I put this private on purpose.

Reply:
Alternatively this could be the primary accessor method ... and the other keeper one could remap over to it.

Reply:
Hmm, I could move it to the keeper.  The other keeper method retrieves record by record id and not scope id and/or name.  I should probably rename this to Records too, since it returns an array.